### PR TITLE
Add generative search

### PR DESCRIPTION
**Pull request by dirkkul**
Needs `semitechnologies/weaviate:preview-add-generative-search-to-grpc-71b2972`

check integration/test_collection_openai.py for examples

### CODE REVIEW

File: weaviate/collection/classes/internal.py [hunk 1]
```diff
@@ -36,6 +37,12 @@ class _Object(Generic[P]):
     metadata: _MetadataReturn
 
 
+@dataclass
+class _GenerativeReturn(Generic[P]):
+    objects: List[_Object[P]]
+    generative_combined_result: Optional[str] = None
```
Comment (tsmith023) [inline weaviate/collection/classes/internal.py:43]: So, to be clear here, `_GenerativeReturn.generative_combined_result` will return the LLM summary of the objects grouped together and `_MetadataReturn.generative` will return the LLM summary of the specific object?

Can users make a generative query and fail to specify `MetadataQuery(generative=True)` and so not retrieve back their LLM generations?  If so, is this something we should default for them?

File: weaviate/collection/grpc.py [hunk 2]
```diff
@@ -376,13 +383,39 @@ def near_audio(
...
+                "Either single_response or grouped response must be not None for generative search."
+            )
+
+        self._generative_single = single
+        self._generative_grouped = grouped
+        self._generative_grouped_properties = grouped_properties
+        self._autocut = auto_limit
+        self._filters = filters
+
         self._metadata = return_metadata
```
Comment (tsmith023) [inline weaviate/collection/grpc.py:412]: Further to my point above, should `MetadataQuery` here have `generative=True` hard-populated `if single is not None`?

**Review by tsmith023 — COMMENTED** [review]
Sorry for the big review on a Friday evening 😭

File: weaviate/collection/grpc.py (hunk 2 above)
Comment (dirkkul) [reply weaviate/collection/grpc.py:412]: not needed (also not possible, there is no generative in MetadataQuery). The server detects if it is a generative query and includes it in the response automatically

File: weaviate/collection/classes/internal.py (hunk 1 above)
Comment (dirkkul) [reply weaviate/collection/classes/internal.py:43]: see below, `MetadataQuery(generative=True) ` does not exist and cannot be forgotten - the reply is send automatically if a generative query is present

File: weaviate/collection/classes/config.py [hunk 3]
```diff
@@ -440,17 +458,28 @@ def to_dict(self) -> Dict[str, Any]:
...
             else:
                 assert isinstance(val, ConfigCreateModel)
                 ret_dict[cls_field] = val.to_dict()
         if self.moduleConfig is None:
             ret_dict["vectorizer"] = Vectorizer.NONE.value
         return ret_dict
 
+    @staticmethod
+    def __add_to_module_config(
+        return_dict: Dict[str, Any], addition_key: str, addition_val: Any
```
[the remark this answers is not in the document]
Comment (dirkkul) [reply weaviate/collection/classes/config.py]: changed it, we can always change it back

File: weaviate/collection/classes/config.py [hunk 4]
```diff
@@ -257,6 +267,13 @@ class PropertyVectorizerConfig(ConfigCreateModel):
     vectorizePropertyName: bool = Field(default=True, alias="vectorize_property_name")
 
 
+class GenerativeFactory:
+    @classmethod
+    def OpenAI(cls) -> GenerativeConfig:
+        """Return a `VectorizerConfig` object with the vectorizer set to `Vectorizer.NONE`"""
```
[the remark this answers is not in the document]
Comment (dirkkul) [reply weaviate/collection/classes/config.py:273]: I have a follow up PR coming were I changed it and added the other options, ok if I leave it in there?

File: weaviate/collection/grpc.py (hunk 2 above)
Comment (tsmith023) [reply weaviate/collection/grpc.py:412]: got confused with `_MetadataReturn` 😅

**Review by tsmith023 — APPROVED** [review]

**Pull request merged**